packages/next-swc/crates/next-core/src/next_client/context.rs

```diff
@@ -65,27 +65,48 @@ use crate::{
     util::foreign_code_context_condition,
 };
 
-fn defines(mode: NextMode) -> CompileTimeDefines {
-    compile_time_defines!(
-        process.turbopack = true,
-        process.env.NODE_ENV = mode.node_env(),
-        process.env.__NEXT_CLIENT_ROUTER_FILTER_ENABLED = false,
-        process.env.__NEXT_HAS_REWRITES = true,
-        process.env.__NEXT_I18N_SUPPORT = false,
-    )
+fn defines(mode: NextMode, dist_root_path: Option<String>) -> CompileTimeDefines {
+    // [TODO] macro may need to allow dynamically expand from some iterable values
+    if let Some(dist_root_path) = dist_root_path {
+        compile_time_defines!(
+            process.turbopack = true,
+            process.env.NODE_ENV = mode.node_env(),
+            process.env.__NEXT_CLIENT_ROUTER_FILTER_ENABLED = false,
+            process.env.__NEXT_HAS_REWRITES = true,
+            process.env.__NEXT_I18N_SUPPORT = false,
+            process.env.__NEXT_DIST_DIR = dist_root_path
+        )
+    } else {
+        compile_time_defines!(
+            process.turbopack = true,
+            process.env.NODE_ENV = mode.node_env(),
+            process.env.__NEXT_CLIENT_ROUTER_FILTER_ENABLED = false,
+            process.env.__NEXT_HAS_REWRITES = true,
+            process.env.__NEXT_I18N_SUPPORT = false,
+        )
+    }
```

Comment:
You can mutate the defines after creating it.
```suggestion
    let mut ctd = compile_time_defines!(
        process.turbopack = true,
        process.env.NODE_ENV = "development",
        process.env.__NEXT_CLIENT_ROUTER_FILTER_ENABLED = false,
        process.env.NEXT_RUNTIME = "edge",
    );
    if let Some(dist_root_path) = dist_root_path {
        ctd.insert(vec!["process".to_string(), "env".to_string(), "__NEXT_DIST_DIR".to_string()], dist_root_path.to_string());
    }
    ctd
```